File: dyros_robot_menagerie/dyros_robot_menagerie/fr3/controller.py
import math
from typing import Dict, Any
import numpy as np
from scipy.spatial.transform import Rotation as R
from rclpy.node import Node
from std_msgs.msg import Int32
from geometry_msgs.msg import PoseStamped
from sensor_msgs.msg import JointState
from mujoco_ros_sim import ControllerInterface
from dyros_robot_menagerie.fr3.robot_data import (FR3RobotData,
                                                  JOINT_DOF,
                                                  TASK_DOF,
                                                  )
from dyros_robot_controller.robot_controller import ManipulatorControllerBase

"""
FR3 MuJoCo Joint/Sensor Information
id | name                 | type   | nq | nv | idx_q | idx_v
----+----------------------+--------+----+----+-------+------
0 | fr3_joint1           | _Hinge |  1 |  1 |     0 |    0
1 | fr3_joint2           | _Hinge |  1 |  1 |     1 |    1
2 | fr3_joint3           | _Hinge |  1 |  1 |     2 |    2
3 | fr3_joint4           | _Hinge |  1 |  1 |     3 |    3
4 | fr3_joint5           | _Hinge |  1 |  1 |     4 |    4
5 | fr3_joint6           | _Hinge |  1 |  1 |     5 |    5
6 | fr3_joint7           | _Hinge |  1 |  1 |     6 |    6

id | name                 | trn     | target_joint
----+----------------------+---------+-------------
0 | fr3_joint1           | _Joint  | fr3_joint1
1 | fr3_joint2           | _Joint  | fr3_joint2
2 | fr3_joint3           | _Joint  | fr3_joint3
3 | fr3_joint4           | _Joint  | fr3_joint4
4 | fr3_joint5           | _Joint  | fr3_joint5
5 | fr3_joint6           | _Joint  | fr3_joint6
6 | fr3_joint7           | _Joint  | fr3_joint7

id | name                        | type             | dim | adr | target (obj)
----+-----------------------------+------------------+-----+-----+----------------
"""
class FR3ControllerPy(ControllerInterface):
    def __init__(self, node: Node, dt: float, mj_joint_dict: Dict[str, Any], ) -> None:
        super().__init__(node, dt, mj_joint_dict)

        self.robot_data       = FR3RobotData(verbose=True)
        self.robot_controller = ManipulatorControllerBase(dt, self.robot_data)
        
        ns = "fr3_controller"
        self._key_sub = node.create_subscription(Int32, f"{ns}/mode_input", self._key_cb, 10)
        self._target_pose_sub = node.create_subscription(PoseStamped, f"{ns}/target_pose", self._target_pose_cb, 10)
        
        self._ee_pose_pub = node.create_publisher(PoseStamped, f"{ns}/ee_pose", 10)
        
        self.mode: str = "NONE"
        self.is_mode_changed = True
        self.is_goal_pose_changed = False

        self.current_time: float = 0.0
        self.control_start_time: float = 0.0

        self.q            = np.zeros(JOINT_DOF)
        self.q_desired    = np.zeros(JOINT_DOF)
        self.q_init       = np.zeros(JOINT_DOF)
        self.qdot         = np.zeros(JOINT_DOF)
        self.qdot_desired = np.zeros(JOINT_DOF)
        self.qdot_init    = np.zeros(JOINT_DOF)

        self.x            = np.eye(4)
        self.x_init       = np.eye(4)
        self.x_desired    = np.eye(4)
        self.xdot         = np.zeros(TASK_DOF)
        self.xdot_init    = np.zeros(TASK_DOF)
        self.xdot_desired = np.zeros(TASK_DOF)

        self.x_goal = np.eye(4)

        self.torque_desired = np.zeros(JOINT_DOF)

    def starting(self) -> None:
        self._ee_timer = self.node.create_timer(0.1, self._pub_ee_pose_cb)

    def updateState(self,
                    pos_dict: Dict[str, np.ndarray],
                    vel_dict: Dict[str, np.ndarray],
                    tau_ext_dict: Dict[str, np.ndarray],
                    sensor_dict: Dict[str, np.ndarray],
                    current_time: float,
                    ) -> None:
        self.current_time = current_time

        # get manipulator joint
        for i in range(JOINT_DOF):
            name = f"fr3_joint{i + 1}"
            self.q[i]    = pos_dict[name][0]
            self.qdot[i] = vel_dict[name][0]

        if not self.robot_data.update_state(self.q, self.qdot):
            self.node.get_logger().error("[FR3RobotData] Failed to update robot state.")

        # get ee
        self.x     = self.robot_data.get_pose()
        self.x_dot = self.robot_data.get_velocity()

    def compute(self) -> None:
        if self.is_mode_changed:
            self.is_mode_changed = False
            self.control_start_time = self.current_time

            self.q_init          = self.q.copy()
            self.qdot_init       = self.qdot.copy()
            self.q_desired       = self.q_init.copy()
            self.qdot_desired[:] = 0.0

            self.x_init          = self.x.copy()
            self.xdot_init       = self.x_dot.copy()
            self.x_desired       = self.x_init.copy()
            self.xdot_desired[:] = 0.0
            
            self.x_goal  = self.x_init.copy()

        if self.mode == "HOME":
            target_q = np.array([0.0, 0.0, 0.0, -math.pi / 2.0, 0.0, math.pi / 2.0, math.pi / 4.0])
            self.torque_desired = self.robot_controller.move_joint_torque_cubic(q_target     = target_q,
                                                                                qdot_target  = np.zeros(JOINT_DOF),
                                                                                q_init       = self.q_init,
                                                                                qdot_init    = self.qdot_init,
                                                                                current_time = self.current_time,
                                                                                init_time    = self.control_start_time,
                                                                                duration     = 4.0,
                                                                                )

        elif self.mode in ("CLIK", "QPIK", "OSF", "QPID"):
            if self.is_goal_pose_changed:
                self.is_goal_pose_changed = False
                self.control_start_time = self.current_time
                
                self.x_init  = self.x.copy()
                self.xdot_init = self.x_dot.copy()

            if self.mode == "CLIK":
                self.qdot_desired = self.robot_controller.CLIK_cubic(x_target     = self.x_goal,
                                                                     xdot_target  = np.zeros(TASK_DOF),
                                                                     x_init       = self.x_init,
                                                                     xdot_init    = self.xdot_init,
                                                                     current_time = self.current_time,
                                                                     init_time    = self.control_start_time,
                                                                     duration     = 4.0,
                                                                     link_name    = self.robot_data.ee_name,
                                                                     )
                self.q_desired += self.dt * self.qdot_desired
                self.torque_desired = self.robot_controller.move_joint_torque_step(self.q_desired, 
                                                                                   self.qdot_desired)

            elif self.mode == "QPIK":
                self.qdot_desired = self.robot_controller.QPIK_cubic(x_target     = self.x_goal,
                                                                     xdot_target  = np.zeros(TASK_DOF),
                                                                     x_init       = self.x_init,
                                                                     xdot_init    = self.xdot_init,
                                                                     current_time = self.current_time,
                                                                     init_time    = self.control_start_time,
                                                                     duration     = 4.0,
                                                                     link_name    = self.robot_data.ee_name,
                                                                     )
                self.q_desired += self.dt * self.qdot_desired
                self.torque_desired = self.robot_controller.move_joint_torque_step(self.q_desired, 
                                                                                   self.qdot_desired)

            elif self.mode == "OSF":
                null_torque = self.robot_controller.move_joint_torque_step(self.q_init, 
                                                                           np.zeros(JOINT_DOF))
                self.torque_desired = self.robot_controller.OSF_cubic(x_target     = self.x_goal,
                                                                      xdot_target  = np.zeros(TASK_DOF),
                                                                      x_init       = self.x_init,
                                                                      xdot_init    = self.xdot_init,
                                                                      current_time = self.current_time,
                                                                      init_time    = self.control_start_time,
                                                                      duration     = 4.0,
                                                                      link_name    = self.robot_data.ee_name,
                                                                      null_torque  = null_torque
                                                                      )

            elif self.mode == "QPID":
                self.torque_desired = self.robot_controller.QPID_cubic(x_target     = self.x_goal,
                                                                      xdot_target  = np.zeros(TASK_DOF),
                                                                      x_init       = self.x_init,
                                                                      xdot_init    = self.xdot_init,
                                                                      current_time = self.current_time,
                                                                      init_time    = self.control_start_time,
                                                                      duration     = 4.0,
                                                                      link_name    = self.robot_data.ee_name,
                                                                      )

        else:
            self.torque_desired = self.robot_data.get_gravity()

    def getCtrlInput(self) -> Dict[str, float]:
        return {f"fr3_joint{i + 1}": float(self.torque_desired[i]) for i in range(JOINT_DOF)}

    def _set_mode(self, mode: str) -> None:
        self.mode = mode
        self.is_mode_changed = True
        self.node.get_logger().info(f"Mode changed: {mode}")

    def _key_cb(self, msg: Int32) -> None:
        mapping = {1: "HOME",
                   2: "CLIK",
                   3: "QPIK",
                   4: "OSF",
                   5: "QPID",
                   }
        self._set_mode(mapping.get(msg.data, "NONE"))

    def _target_pose_cb(self, msg: PoseStamped) -> None:
        self.is_goal_pose_changed = True

        quat = np.array([msg.pose.orientation.x,
                         msg.pose.orientation.y,
                         msg.pose.orientation.z,
                         msg.pose.orientation.w,
                         ])
        rot = R.from_quat(quat).as_matrix()
        self.x_goal = np.eye(4)
        self.x_goal[:3, :3] = rot
        self.x_goal[:3, 3] = [msg.pose.position.x, msg.pose.position.y, msg.pose.position.z]

    def _pub_ee_pose_cb(self) -> None:
        msg = PoseStamped()
        msg.header.frame_id = "base_link"
        msg.header.stamp = self.node.get_clock().now().to_msg()

        msg.pose.position.x, msg.pose.position.y, msg.pose.position.z = self.x[:3, 3]
        quat = R.from_matrix(self.x[:3, :3]).as_quat()  # x, y, z, w
        msg.pose.orientation.x, msg.pose.orientation.y, msg.pose.orientation.z, msg.pose.orientation.w = quat
        self._ee_pose_pub.publish(msg)

File: dyros_robot_menagerie/src/fr3/controller.cpp
#include "fr3/controller.h"

namespace FR3
{
    FR3Controller::FR3Controller(const rclcpp::Node::SharedPtr& node)
    : ControllerInterface(node)
    {
        dt_ = 0.001;

        robot_data_ = std::make_shared<FR3RobotData>();
        robot_controller_ = std::make_unique<RobotController::Manipulator::ManipulatorBase>(dt_, robot_data_);
        
        key_sub_ = node_->create_subscription<std_msgs::msg::Int32>("fr3_controller/mode_input", 10,std::bind(&FR3Controller::keyCallback, this, std::placeholders::_1));
        target_pose_sub_ = node_->create_subscription<geometry_msgs::msg::PoseStamped>("fr3_controller/target_pose", 10,std::bind(&FR3Controller::subtargetPoseCallback, this, std::placeholders::_1));
        
        ee_pose_pub_ = node_->create_publisher<geometry_msgs::msg::PoseStamped>("fr3_controller/ee_pose", 10);
        hand_eye_cam_pub_ = node_->create_publisher<sensor_msgs::msg::Image>("fr3_controller/hand_eye_cam", 10);
        
        q_.setZero();
        q_desired_.setZero();
        q_init_.setZero();
        qdot_.setZero();
        qdot_desired_.setZero();
        qdot_init_.setZero();
        
        x_.setIdentity();
        x_init_.setIdentity();
        x_desired_.setIdentity();
        xdot_.setZero();
        xdot_init_.setZero();
        xdot_desired_.setZero();
        
        x_goal_.setIdentity();
        
        torque_desired_.setZero();
        
        std::ostringstream oss;
        oss << "\n=================================================================\n"
            << "=================================================================\n"
            << "URDF Joint Information: FR3\n"
            << robot_data_->getVerbose()
            << "=================================================================\n"
            << "=================================================================";
        const std::string print_info = oss.str();
        RCLCPP_INFO(node->get_logger(), "%s%s%s", cblue, print_info.c_str(), creset);
    }

    FR3Controller::~FR3Controller()
    {

    }
    
    void FR3Controller::starting()
    {
        ee_pose_pub_timer_ = node_->create_wall_timer(std::chrono::milliseconds(100), std::bind(&FR3Controller::pubEEPoseCallback, this));
        hand_eye_cam_pub_timer_ = node_->create_wall_timer(std::chrono::milliseconds(33), std::bind(&FR3Controller::pubHandEyeCallback, this));
    }

    void FR3Controller::updateState(const VecMap& pos_dict, 
                                    const VecMap& vel_dict,
                                    const VecMap& tau_ext_dict, 
                                    const VecMap& sensors_dict, 
                                    double current_time)
    {
        current_time_ = current_time;

        // get manipulator joint
        for(size_t i=0; i<JOINT_DOF; i++)
        {
            const std::string& name = "fr3_joint" + std::to_string(i+1);
            q_(i) = pos_dict.at(name)(0);
            qdot_(i) = vel_dict.at(name)(0);
        }

        if(!robot_data_->updateState(q_, qdot_)) RCLCPP_ERROR(node_->get_logger(), "%sFailed to update robot state.%s", cred, creset);

        // get ee
        x_ = robot_data_->getPose();
        xdot_ = robot_data_->getVelocity();
    }

    void FR3Controller::updateRGBDImage(const ImageCVMap& images)
    {
        std::scoped_lock<std::mutex> lk(hand_eye_cam_mtx_);
        hand_eye_cam_img_ = images.at("hand_eye").clone();

        if (hand_eye_cam_img_.type() == CV_32FC1)     hand_eye_cam_encoding_ = "32FC1";
        else if (hand_eye_cam_img_.type() == CV_8UC1) hand_eye_cam_encoding_ = "mono8";
        else                                          hand_eye_cam_encoding_ = "rgb8";
    }

    void FR3Controller::compute()
    {
        if(is_mode_changed_)
        {
            is_mode_changed_ = false;
            control_start_time_ = current_time_;

            q_init_ = q_;
            qdot_init_ = qdot_;
            q_desired_ = q_init_;
            qdot_desired_.setZero();

            x_init_ = x_;
            xdot_init_ = xdot_;
            x_desired_ = x_init_;
            xdot_desired_.setZero();

            x_goal_ = x_init_;
        }

        if(mode_ == "HOME")
        {
            JointVec target_q;
            target_q << 0.0, 0.0, 0.0, -M_PI/2., 0.0, M_PI/2., M_PI / 4.;
            torque_desired_ = robot_controller_->moveJointTorqueCubic(target_q,
                                                                      JointVec::Zero(),
                                                                      q_init_,
                                                                      qdot_init_,
                                                                      current_time_,
                                                                      control_start_time_,
                                                                      4.0);
        }
        else if(mode_ == "CLIK")
        {
            if(is_goal_pose_changed_)
            {
                control_start_time_ = current_time_;
                x_init_ = x_;
                xdot_init_ = xdot_;
                is_goal_pose_changed_ = false;
            }
            qdot_desired_ = robot_controller_->CLIKCubic(x_goal_,
                                                         JointVec::Zero(),
                                                         x_init_,
                                                         xdot_init_,
                                                         current_time_,
                                                         control_start_time_,
                                                         4.0,
                                                         robot_data_->getEEName());
            q_desired_ += dt_ * qdot_desired_;
            torque_desired_ = robot_controller_->moveJointTorqueStep(q_desired_, qdot_desired_);
        }
        else if(mode_ == "QPIK")
        {
            if(is_goal_pose_changed_)
            {
                control_start_time_ = current_time_;
                x_init_ = x_;
                xdot_init_ = xdot_;
                is_goal_pose_changed_ = false;
            }
            qdot_desired_ = robot_controller_->QPIKCubic(x_goal_,
                                                         JointVec::Zero(),
                                                         x_init_,
                                                         xdot_init_,
                                                         current_time_,
                                                         control_start_time_,
                                                         4.0,
                                                         robot_data_->getEEName());
            q_desired_ += dt_ * qdot_desired_;
            torque_desired_ = robot_controller_->moveJointTorqueStep(q_desired_, qdot_desired_);
        }
        else if(mode_ == "OSF")
        {
            if(is_goal_pose_changed_)
            {
                control_start_time_ = current_time_;
                x_init_ = x_;
                xdot_init_ = xdot_;
                is_goal_pose_changed_ = false;
            }
            JointVec target_q;
            target_q << 0.0, 0.0, 0.0, -M_PI/2., 0.0, M_PI/2., M_PI / 4.;
            VectorXd null_torque_desired = robot_controller_->moveJointTorqueStep(q_init_, JointVec::Zero());
            torque_desired_ = robot_controller_->OSFCubic(x_goal_,
                                                          JointVec::Zero(),
                                                          x_init_,
                                                          xdot_init_,
                                                          current_time_,
                                                          control_start_time_,
                                                          4.0,
                                                          null_torque_desired,
                                                          robot_data_->getEEName());
        }
        else if(mode_ == "QPID")
        {
            if(is_goal_pose_changed_)
            {
                control_start_time_ = current_time_;
                x_init_ = x_;
                xdot_init_ = xdot_;
                is_goal_pose_changed_ = false;
            }
            torque_desired_ = robot_controller_->QPIDCubic(x_goal_,
                                                           JointVec::Zero(),
                                                           x_init_,
                                                           xdot_init_,
                                                           current_time_,
                                                           control_start_time_,
                                                           4.0,
                                                           robot_data_->getEEName());
        }
        else
        {
            torque_desired_ = robot_data_->getGravity();
        }
    }

    CtrlInputMap FR3Controller::getCtrlInput() const
    {
        CtrlInputMap ctrl_dict;
        for(size_t i=0; i<JOINT_DOF; i++)
        {
            const std::string name = "fr3_joint" + std::to_string(i+1);
            ctrl_dict[name] = torque_desired_(i);
        }
        return ctrl_dict;
    }

    void FR3Controller::setMode(const std::string& mode)
    {
        is_mode_changed_ = true;
        mode_ = mode;
        RCLCPP_INFO(node_->get_logger(), "%sMode changed: %s%s", cblue, mode.c_str(), creset);
    }

    void FR3Controller::keyCallback(const std_msgs::msg::Int32::SharedPtr msg)
    {
        RCLCPP_INFO(node_->get_logger(), "Key input received: %d", msg->data);
        if(msg->data == 1)      setMode("HOME");
        else if(msg->data == 2) setMode("CLIK");
        else if(msg->data == 3) setMode("QPIK");
        else if(msg->data == 4) setMode("OSF");
        else if(msg->data == 5) setMode("QPID");
        else                    setMode("NONE");
    }

    void FR3Controller::subtargetPoseCallback(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
    {
        RCLCPP_INFO(node_->get_logger(),
                    "Target pose received: position=(%.3f, %.3f, %.3f), "
                    "orientation=(%.3f, %.3f, %.3f, %.3f)",
                    msg->pose.position.x, msg->pose.position.y, msg->pose.position.z,
                    msg->pose.orientation.x, msg->pose.orientation.y,
                    msg->pose.orientation.z, msg->pose.orientation.w);
        is_goal_pose_changed_ = true;

        // Convert to 4x4 homogeneous transform
        Eigen::Quaterniond quat(msg->pose.orientation.w,
                                msg->pose.orientation.x,
                                msg->pose.orientation.y,
                                msg->pose.orientation.z);

        x_goal_.linear() = quat.toRotationMatrix();
        x_goal_.translation() << msg->pose.position.x,
                                msg->pose.position.y,
                                msg->pose.position.z;
    }

    void FR3Controller::pubEEPoseCallback()
    {
        auto ee_pose_msg = geometry_msgs::msg::PoseStamped();
        ee_pose_msg.header.frame_id = "base_link";
        ee_pose_msg.header.stamp = node_->now();

        ee_pose_msg.pose.position.x = x_.translation()(0);
        ee_pose_msg.pose.position.y = x_.translation()(1);
        ee_pose_msg.pose.position.z = x_.translation()(2);

        Eigen::Quaterniond q(x_.rotation());
        ee_pose_msg.pose.orientation.x = q.x();
        ee_pose_msg.pose.orientation.y = q.y();
        ee_pose_msg.pose.orientation.z = q.z();
        ee_pose_msg.pose.orientation.w = q.w();
        
        ee_pose_pub_->publish(ee_pose_msg);
    }

    void FR3Controller::pubHandEyeCallback()
    {
        cv::Mat img;
        std::string enc;
        {
            std::scoped_lock<std::mutex> lk(hand_eye_cam_mtx_);
            if (hand_eye_cam_img_.empty()) return;
            img = hand_eye_cam_img_.clone();
            enc = hand_eye_cam_encoding_;
        }

        auto msg = toImageMsg(img, enc);
        msg->header.stamp = node_->now();
        msg->header.frame_id = "hand_eye_cam_frame";
        hand_eye_cam_pub_->publish(*msg);
    }
    
    /* register with the global registry */
    PLUGINLIB_EXPORT_CLASS(FR3::FR3ControllerFactory, ControllerFactory)
}